keps/sig-cli/2953-kustomize-plugin-graduation/README.md

```diff
@@ -0,0 +1,764 @@
+
+This KEP proposes using the concept of a "trusted plugin catalog" from the [Catalog KEP] as a convenient, concise way for end users to express their intent to execute a constrained set of plugins. With Catalog, the list of plugins being authorized is explicit and reviewable.
+
+**Plugins from a user-specified trusted catalog**
+
+When using plugins with a trusted catalog, only one flag will be required, regardless of the runtime type of the provider: `--trusted-plugin-catalog=""` .
+
+This new repeatable flag will be introduced by the [Catalog KEP]. When that flag is present, plugins in the referenced catalogs will be executed without the need for any additional flags. The integrity of the plugin provider will be validated against the Catalog entry before execution. This all applies whether the providers are defined explicitly in the Kustomization (and the explicit entry matches the catalog) or looked up dynamically in the catalog.
+
+If a containerized plugin needs network or disk access, its catalog entry MUST specify it. Redundant inline provider config cannot be used to grant it. When the catalog does specify it, no additional flags or config fields will be required to grant it the required access.
+
+**Plugins from a compiled-in trusted catalog**
+
+Optionally, it should be possible for a user to compile a bespoke distribution of Kustomize that embeds a trusted catalog, such that the plugins it references can be used without any additional flags on the end user's part. The target persona for this scenario is a platform maintainer at a large organization that distributes its own Kustomize internally. This approach could also be used in the future to extract Kustomize built-ins to plugins to reduce compiled in dependencies while minimizing UX impact.
+
+**Plugins NOT in a trusted catalog**
+
+*Recommended Option: no such thing*
+
+Trusted catalogs are always required to use plugins; uncatalogued plugins will not be executed. Kustomizations that require ad-hoc plugins must be accompanied by a `Catalog` that the end user can reference locally, e.g. `kustomize build dir/ --trusted-plugin-catalog=dir/catalog.yaml`. There will be no additional plugin-related flags. The end user will be expected to inspect this catalog before trusting it, which must always be done explicitly on the command line.
+
+A `kustomize edit generate-catalog` command will be implemented to streamline local plugin workflows. That command will take a Kustomization, collect all the uncatalogued plugins into a list, and build that list into the catalog resource format, using local references relative to the root Kustomization (plugins outside the root will result in an error). Optionally, this command can extract the provider references from the Kustomization, since they are most likely redundant once the catalog has been constructed. Optionally, this command can localize all required entries from referenced remote catalogs into the generated catalog.
```

Comment:
this does make stuff easier but introduces another layer that needs to be synchronized.  This is akin to the C++ header files that eventually language designers went away from.  My recommendation is to minimize the need for boilerplate even if it's automatically generated.

Reply:
Do you have an alternate suggestion for addressing the security problem that Catalog is being used for here? I don't see Catalog content as boilerplate: on the contrary, it is the centrepiece of this declarative plugin model that enables confidence in plugin execution. The alternative outlined below (pointing to allowed plugins in the flag itself) is more effort for every end user. The existing system of a flag generically allowing any plugin is not adequate for security, especially in a context where remote Kustomizations are used (imagine you intend to allow your own local plugin to run, and the remote Kustomization you're importing changes to contain an exec plugin--IMO it should not silently run). We could add a separate set of flags to generically enable plugins contained within the Kustomization root, but my feeling is that that would add more complexity overall, since users would need to learn the distinction in how to use "local" and non-local plugins instead of having a single consistent model.

Reply:
Clarification: so the idea of having a catalog is to ensure that when I have remote resources that those do not run an unknown plugin on my system?  So the solution then is to ensure that I put in the right list of plugins into the catalog so that the remote resources are only able to execute the plugins that I have put into my trusted catalog.  This does have the benefit that instead of executing something the CD pipeline will break and a human operator will be forced to take a look and verify that the file needs to be updated.  I agree that exec plugins should not silently run as they pose a greater danger to the environment.
I think the issue with remotes is much deeper.  If you are importing a remote resource from a place that is not stable or not trustworthy you might have a totally different set of problems.
If we are adding tax on all the kustomize users to enable remote plugins... I am a bit conflicted by that.  In general I consider the remotes the way they are done right now a recipe for unwelcome surprises.  I would not reinforce those scenarios but instead look to localize.  https://github.com/kubernetes-sigs/kustomize/issues/3980

Reply:
> Clarification: so the idea of having a catalog is to ensure that when I have remote resources that those do not run an unknown plugin on my system? 

Not remote resources specifically, but yes, this KEP proposes leveraging Catalog (which is also proposed for independent reasons related to plugin discovery, distribution and end-user boilerplate reduction) as a trust model, and that trust model role is the reason for the flag, as well as for the concept of a "local" catalog (the paragraph we're commenting on here). 

> So the solution then is to ensure that I put in the right list of plugins into the catalog so that the remote resources are only able to execute the plugins that I have put into my trusted catalog. 

Yes, that's what is proposed here. Kustomize needs a way for end users to tell us that they are aware that third party code X is going to be executed, and they approve of it. Since Catalog is already a way to list plugins from a given source, it feels like a natural fit for this problem. 

> If we are adding tax on all the kustomize users to enable remote plugins... I am a bit conflicted by that.

Plugins from remote Kustomizations are the worst-case scenario to consider in evaluating options, but we need end-user "approval" in some form in order for `kustomize build` to run ANY third-party code, regardless of the Kustomization or the plugin being remote or local.

Here's an idea that we could implement if end-users complain about the flag being burdensome: since it is end-user trust we need (i.e. the person running `kustomize build`, not necessarily the author of that command's target Kustomization), we could create a user config file (e.g. `$XDG_DATA_HOME/kustomize/config`) containing settings like a list of catalogs that user wants to trust by default. So there would be three trusted catalog sources: the binary itself (for the hypothetical compiled-in catalog), the flag, and the config file. Wdyt?